Fix Issue 2023: Validate VolumeClaimTemplate overrides contain `spec` and provide helpful error messages when they don't

jmealo opened the pull request:
**This PR is intended to fix #2023.**

  **Note to reviewers:** remember to look at the commits in this PR and consider if they can be squashed

  ## Summary Of Changes

  This PR adds validation and improved error messages to prevent silent reconciliation failures when users provide incomplete `volumeClaimTemplates` in the `override.statefulSet.spec` section.

  ### Problem
  When users add only `metadata.annotations` to `volumeClaimTemplates` (e.g., for PVC autoresizer annotations) without including the required `spec.resources.requests.storage` field, the override replaces the entire template
  rather than merging. This results in:
  - PVC templates with `storage: 0` (missing field defaults to zero)
  - Silent reconciliation loop failures every ~15 minutes
  - No clear indication to users about what's wrong
  - Cluster appears to be running but changes don't apply

  ### Changes

  1. **CRD Schema Validation** (`api/v1beta1/rabbitmqcluster_types.go`)
     - Added `+kubebuilder:validation:Required` marker to `PersistentVolumeClaim.Spec` field
     - Kubernetes API server now rejects incomplete PVC templates at admission time
     - Users get immediate feedback: `spec.override.statefulSet.spec.volumeClaimTemplates[0].spec: Required value`

  2. **Improved Error Messages** (`controllers/reconcile_persistence.go`)
     - Detects when PVC template has `storage=0` or missing storage field
     - Returns helpful error explaining that overrides replace entire templates (don't merge)
     - Clarifies that complete `spec.resources.requests.storage` must be provided

  3. **Enhanced Shrink Error** (`internal/scaling/scaling.go`)
     - Shows actual capacity values in error message: `"shrinking not supported (existing: 20Gi, desired: 5Gi)"`
     - Makes troubleshooting capacity issues more intuitive

  4. **Regenerated CRD** (`config/crd/bases/rabbitmq.com_rabbitmqclusters.yaml`)
     - Updated with new validation requirements

  ### What This Does NOT Change
  - No changes to comparison logic (still uses `.Cmp()`)
  - No changes to test files
  - No behavioral changes to valid configurations
  - Ephemeral storage (`storage: "0"`) continues to work correctly

  ## Additional Context

  **Root Cause:** The issue typically manifests when users (or LLM tools) treat the override as a merge instead of a replace (and leave out the `spec` portion, causing `size` to default to `0`)

  ```yaml
  override:
    statefulSet:
      spec:
        volumeClaimTemplates:
          - metadata:
              annotations:
                resize.topolvm.io/storage_limit: 100Gi
            # Missing spec.resources.requests.storage!
```
Since `override.statefulSet.spec.volumeClaimTemplates` is a full replacement (not a strategic merge), the above results in a PVC template with no storage capacity, which defaults to `0`.

  **After This PR:**
  - Invalid configs are rejected immediately by the API server
  - If somehow they get through, reconciliation fails fast with a clear error
  - Ephemeral storage continues to work (tested)
  - Works as intended when deployed with latest operator and CRDs
  - Works as intended when only latest CRDs are deployed (without operator update)
  - Works as intended (minus immediate feedback) when only latest operator is deployed (without latest CRDS)

# Manual Testing Scenarios
These all passed for me on a fresh Azure AKS cluster that did not have RabbitMQ or the operator deployed.

## ✅ Test 1: Invalid override is rejected
```kubectl apply -f - <<'EOF'
apiVersion: rabbitmq.com/v1beta1
kind: RabbitmqCluster
metadata:
  name: test-invalid
spec:
  replicas: 1
  override:
    statefulSet:
      spec:
        volumeClaimTemplates:
          - metadata:
              name: persistence
              annotations:
                resize.topolvm.io/storage_limit: 100Gi
EOF
The RabbitmqCluster "test-invalid" is invalid: spec.override.statefulSet.spec.volumeClaimTemplates[0].spec: Required value
```

### Expected: Rejected with "spec: Required value"

## ✅ Test 2: Valid ephemeral cluster still works
```kubectl apply -f - <<EOF
apiVersion: rabbitmq.com/v1beta1
kind: RabbitmqCluster
metadata:
  name: test-ephemeral
spec:
  replicas: 1
  persistence:
    storage: "0"
EOF
rabbitmqcluster.rabbitmq.com/test-ephemeral created
```
### Expected: Cluster created successfully, no PVCs

## ✅ Test 3: Valid override with complete spec
```kubectl apply -f - <<EOF
apiVersion: rabbitmq.com/v1beta1
kind: RabbitmqCluster
metadata:
  name: test-valid
spec:
  replicas: 1
  override:
    statefulSet:
      spec:
        volumeClaimTemplates:
          - metadata:
              name: persistence
              annotations:
                resize.topolvm.io/storage_limit: 100Gi
            spec:
              accessModes: [ReadWriteOnce]
              resources:
                requests:
                  storage: 20Gi
EOF
rabbitmqcluster.rabbitmq.com/test-valid created
```
### Expected: Cluster created successfully with 20Gi PVCs

# Test Results
  - ✅ All unit tests pass (347 specs)
  - ✅ All integration tests pass (59 specs)
  - ✅ Ephemeral storage clusters work correctly
  - ✅ Invalid configs are rejected at admission time
  - ✅ Improved error messages appear in operator logs when applicable

jmealo commented:
**I tested this and it works:**
```
Failed to scale PVCs: shrinking persistent volumes is not supported (existing: 20Gi, desired: 0)
```

Not sure why the desired size is: 0, but I'm guessing a regression in our helm chart. The logging improvements on this PR made this a lot easier to figure out though. Still troubleshooting the actual issue.

Zerpet commented:
> **I tested this and it works:**
> 
> ```
> Failed to scale PVCs: shrinking persistent volumes is not supported (existing: 20Gi, desired: 0)
> ```
> 
> Not sure why the desired size is: 0, but I'm guessing a regression in our helm chart. The logging improvements on this PR made this a lot easier to figure out though. Still troubleshooting the actual issue.

This is the function that gets the desired capacity, and all it does is grab it from the rabbitmq cluster spec:

https://github.com/rabbitmq/cluster-operator/blob/main/controllers/reconcile_persistence.go#L15-L18

I'm not opposed to adding more logging at higher verbosity level, if it's helpful. However, replacing `Cmp()` to do exactly what `Cmp()` does it's very unlikely to be accepted as contribution:

https://github.com/kubernetes/apimachinery/blob/b72d93d174332f952a8d431419fece5e6f044bcb/pkg/api/resource/quantity.go#L640-L645

jmealo commented:
@Zerpet: Thanks! I completely agree, and I saw the docs for `cmp` said they should already be doing this. To say I was surprised about the desired size of 0 is an understatement. I'm digging in on this today, checking my other clusters for the issue. So far nothing in the `helm` output is suspect. I think adding non-verbose logging for the desired size of 0 would be worth doing to save others the trouble (depending on what the root cause is). I've time boxed it for an hour to try to figure it out 🤞. **Thank you for the quick review! 🙏**

jmealo committed:
Fix PVC storage validation and error messages

Prevents silent reconciliation failures when override.statefulSet.spec.volumeClaimTemplates
is provided with incomplete configuration (e.g., only metadata.annotations without
spec.resources.requests.storage).

Changes:
- Add CRD validation requiring PVC spec field (rejected at admission time)
- Detect and error on storage=0 with hints about override behavior
- Show actual values in shrink errors: "(existing: 20Gi, desired: 5Gi)"

Fixes #2023

jmealo committed:
Apply suggestions from code review

Co-authored-by: Aitor Pérez Cedres <1515757+Zerpet@users.noreply.github.com>

jmealo commented:
@Zerpet Suggestions applied. Thanks for reviewing!

Zerpet commented:
This PR has gone radio silent for over 90 days. Closing for now.
